File: trunk/Gomoku/Client/cpp/Gomoku/Core/Source/WebClient.cpp
/*
 *  Gomoku Core
 * 
 *  Copyright (c) 2011 Tran Dinh Thoai <dthoai@yahoo.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "..\Header\WebClient.h"
#include "..\Private\Socket.h" // Requires Ws2_32.lib, see: http://www.adp-gmbh.ch/win/misc/sockets.html
#include "..\Header\Board.h"

using namespace gomoku_core;

WebClient::WebClient(Config *config) : Client(config) {
	this->ready = false;
}

WebClient::~WebClient() {

}

string *WebClient::request(string link) {
	string host;
	string path;
	size_t pos = link.find("//");
	if (pos == string::npos) {
		host = link.substr(0);
	} else {
		host = link.substr(pos + 2);
	}
	pos = host.find("/");
	if (pos == string::npos) {
		path = "";
	} else {
		path = host.substr(pos);
		host = host.substr(0, pos);
	}
	string *tag = new string("");
	try {
		SocketClient s(host, 80);
		s.SendLine("GET " + path + " HTTP/1.0");
		s.SendLine("Host: " + host);
		s.SendLine("");

		bool start = false;
		while (true) {
			string line = s.ReceiveLine();
			if (line.empty()) break;
			if (start) {
				tag->append(line);
			} else {
				if (line == "\r\n") {
					start = true;
					continue;
				}
			}
		}
	} catch (const char* s) {
		*tag = "Error: ";
		tag->append(s);
	} catch (string s) {
		*tag = "Error: ";
		tag->append(s);
	} catch (...) {
		*tag = "Error: Unhandled exception";
	}

	return tag;
}

void WebClient::login() {
	if (this->ready) {
		this->logout();
	}
		
	string *serverUrl = this->config->getString(Config::SERVER_URL);
	string *extension = this->config->getString(Config::SERVER_EXTENSION);
	string *username = this->config->getString(Config::SERVER_USERNAME);
	string *password = this->config->getPassword(Config::SERVER_PASSWORD);
	char buffer[255];
	sprintf(buffer, "%sapi/login%s?u=%s&p=%s", serverUrl->c_str(), extension->c_str(), username->c_str(), password->c_str());
	string *link = new string(buffer);
	string *response = request(*link);
	delete serverUrl;
	delete extension;
	delete username;
	delete password;
	delete link;
	string sign = "Success: ";
	if (response->find(sign) == 0) {
		this->session = response->substr(sign.length());
		this->ready = true;
		delete response;
		return;
	}
	sign = "Error: ";
	if (response->find(sign) == 0) {
		string message = "Login fail: " + response->substr(sign.length());
		delete response;
		throw message;
	}
	delete response;
	throw string("Login fail: Invalid api call!");
}

void WebClient::logout() {
	try {
		string *serverUrl = this->config->getString(Config::SERVER_URL);
		string *extension = this->config->getString(Config::SERVER_EXTENSION);
		char buffer[255];
		sprintf(buffer, "%sapi/logout%s?s=%s", serverUrl->c_str(), extension->c_str(), this->session.c_str());
		string *link = new string(buffer);
		string *response = request(*link);
		delete serverUrl;
		delete extension;
		delete link;
		delete response;
	} catch (...) { }
	this->ready = false;
	this->session = "";
}

bool WebClient::online() {
	return this->ready;
}

void WebClient::clone(string session) {
	this->session = session;
	this->ready = true;
}

string *WebClient::createGame(bool playFirst, int width, int height) {
	if (!this->ready) {
		this->login();
	}

	string *serverUrl = this->config->getString(Config::SERVER_URL);
	string *extension = this->config->getString(Config::SERVER_EXTENSION);
	char buffer[255];
	sprintf(buffer, "%sapi/create%s?s=%s&p=%s&w=%d&h=%d", serverUrl->c_str(), extension->c_str(), this->session.c_str(), playFirst ? "1" : "0", width, height);
	string *link = new string(buffer);
	string *response = request(*link);
	delete serverUrl;
	delete extension;
	delete link;
	string sign = "Success: ";
	if (response->find(sign) == 0) {
		string *tag = new string(response->substr(sign.length()));
		delete response;
		return tag;
	}
	sign = "Error: ";
	if (response->find(sign) == 0) {
		string message = "Fail to create game: " + response->substr(sign.length());
		delete response;
		throw message;
	}
	delete response;
	throw string("Fail to create game: Invalid api call!");
}

void WebClient::joinGame(string gameId) {
	if (!this->ready) {
		this->login();
	}

	string *serverUrl = this->config->getString(Config::SERVER_URL);
	string *extension = this->config->getString(Config::SERVER_EXTENSION);
	char buffer[255];
	sprintf(buffer, "%sapi/join%s?s=%s&g=%s", serverUrl->c_str(), extension->c_str(), this->session.c_str(), gameId.c_str());
	string *link = new string(buffer);
	string *response = request(*link);
	delete serverUrl;
	delete extension;
	delete link;
	string sign = "Error: ";
	if (response->find(sign) == 0) {
		string message = "Fail to join game: " + response->substr(sign.length());
		delete response;
		throw message;
	}
	if (*response != "Success") {
		delete response;
		throw string("Fail to join game: Invalid api call!");
	}
	delete response;
}

void WebClient::listGame(list<string> &tag) {
	try {
		tag.clear();

		if (!this->ready) {
			this->login();
		}

		string *serverUrl = this->config->getString(Config::SERVER_URL);
		string *extension = this->config->getString(Config::SERVER_EXTENSION);
		char buffer[255];
		sprintf(buffer, "%sapi/list%s?s=%s", serverUrl->c_str(), extension->c_str(), this->session.c_str());
		string *link = new string(buffer);
		string *response = request(*link);
		delete serverUrl;
		delete extension;
		delete link;
		string sign = "Success: ";
		if (response->find(sign) == 0) {
			string data = response->substr(sign.length());
			if (data.length() > 0) {
				this->split(data, "\r\n", tag);
			}
		}
		delete response;
	} catch (...) { }
}

Move *WebClient::lastMove() {
	try {
		if (!this->ready) {
			this->login();
		}

		string *serverUrl = this->config->getString(Config::SERVER_URL);
		string *extension = this->config->getString(Config::SERVER_EXTENSION);
		string *gameId = this->config->getString(Config::CURRENT_GAME);
		char buffer[255];
		sprintf(buffer, "%sapi/last%s?s=%s&g=%s", serverUrl->c_str(), extension->c_str(), this->session.c_str(), gameId->c_str());
		string *link = new string(buffer);
		string *response = request(*link);
		delete serverUrl;
		delete extension;
		delete gameId;
		delete link;
		string sign = "Success: ";
		if (response->find(sign) == 0) {
			string data = response->substr(sign.length());
			delete response;
			if (data.size() == 0) {
				return new Move();
			} else {
				list<string> fields;
				this->split(data, "|", fields);
				list<string>::iterator it;
				unsigned char piece = Board::BLANK;
				unsigned char row = 0;
				unsigned char column = 0;
				unsigned char no = 0;
				for (it = fields.begin(); it != fields.end(); it++, no++) {
					if (no == 0) {
						piece = unsigned char(atoi((*it).c_str()));
						if (piece != Board::BLACK && piece != Board::WHITE) {
							return new Move();
						}
					}
					if (no == 1) {
						row = unsigned char(atoi((*it).c_str()));
						if (row >= this->config->getByte(Config::BOARD_HEIGHT)) {
							return new Move();
						}
					}
					if (no == 2) {
						column = unsigned char(atoi((*it).c_str()));
						if (column >= this->config->getByte(Config::BOARD_WIDTH)) {
							return new Move();
						}
					}
				}
				return new Move(row, column, piece);
			}
		} else {
			delete response;
			return new Move();
		}
	} catch (...) { 
		return new Move();
	}
}

void WebClient::makeMove(Move move) {
	if (!this->ready) {
		this->login();
	}

	string *serverUrl = this->config->getString(Config::SERVER_URL);
	string *extension = this->config->getString(Config::SERVER_EXTENSION);
	string *gameId = this->config->getString(Config::CURRENT_GAME);
	char buffer[255];
	sprintf(buffer, "%sapi/move%s?s=%s&g=%s&r=%d&c=%d&p=%d", serverUrl->c_str(), extension->c_str(), this->session.c_str(), gameId->c_str(), move.getRow(), move.getColumn(), move.getPiece());
	string *link = new string(buffer);
	string *response = request(*link);
	delete serverUrl;
	delete extension;
	delete gameId;
	delete link;
	string sign = "Error: ";
	if (response->find(sign) == 0) {
		string message = "Fail to make move: " + response->substr(sign.length());
		delete response;
		throw message;
	}
	if (*response != "Success") {
		delete response;
		throw string("Fail to make move: Invalid api call!");
	}
	delete response;
}

void WebClient::gameState(GameState &state) {
	if (!this->ready) {
		this->login();
	}

	string *serverUrl = this->config->getString(Config::SERVER_URL);
	string *extension = this->config->getString(Config::SERVER_EXTENSION);
	string *gameId = this->config->getString(Config::CURRENT_GAME);
	char buffer[255];
	sprintf(buffer, "%sapi/state%s?s=%s&g=%s", serverUrl->c_str(), extension->c_str(), this->session.c_str(), gameId->c_str());
	string *link = new string(buffer);
	string *response = request(*link);
	delete serverUrl;
	delete extension;
	delete gameId;
	delete link;
	string sign = "Success: ";
	if (response->find(sign) == 0) {
		string data = response->substr(sign.length());
		delete response;
		list<string> fields;
		this->split(data, "|", fields);
		list<string>::iterator it;
		unsigned char no = 0;
		for (it = fields.begin(); it != fields.end(); it++, no++) {
			if (no == 0) {
				state.Joined = (*it) == "1";
			}
			if (no == 1) {
				state.Cancelled = (*it) == "1";
			}
			if (no == 2) {
				state.Finished = (*it) == "1";
			}
			if (no == 3) {
				state.Victory = unsigned char(atoi((*it).c_str()));
			}
		}
		return;
	}
	sign = "Error: ";
	if (response->find(sign) == 0) {
		string message = "Fail to get game state: " + response->substr(sign.length());
		delete response;
		throw message;
	}
	delete response;
	throw string("Fail to get game state: Invalid api call!");
}

void WebClient::endGame(bool finished, unsigned char victory) {
	if (!this->ready) {
		this->login();
	}

	string *serverUrl = this->config->getString(Config::SERVER_URL);
	string *extension = this->config->getString(Config::SERVER_EXTENSION);
	string *gameId = this->config->getString(Config::CURRENT_GAME);
	char buffer[255];
	sprintf(buffer, "%sapi/end%s?s=%s&g=%s&f=%s&v=%d", serverUrl->c_str(), extension->c_str(), this->session.c_str(), gameId->c_str(), finished ? "1" : "0", victory);
	string *link = new string(buffer);
	string *response = request(*link);
	delete serverUrl;
	delete extension;
	delete gameId;
	delete link;
	string sign = "Error: ";
	if (response->find(sign) == 0) {
		string message = "Fail to end game: " + response->substr(sign.length());
		delete response;
		throw message;
	}
	if (*response != "Success") {
		delete response;
		throw string("Fail to end game: Invalid api call!");
	}
	delete response;
}

void WebClient::split(string src, string sep, list<string> &tag) {
	tag.clear();
	size_t curPos = 0;
	size_t nextPos = src.find(sep);
	while (nextPos != string::npos) {
		string line = src.substr(curPos, nextPos - curPos);
		tag.push_back(line);
		curPos = nextPos + sep.length();
		nextPos = src.find(sep, curPos);
	}
	tag.push_back(src.substr(curPos));
}

File: Gomoku/Client/java/Core/src/com/bhivef/gomoku/core/WebClient.java
/*
 *  Gomoku Core
 * 
 *  Copyright (c) 2011 Tran Dinh Thoai <dthoai@yahoo.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

package com.bhivef.gomoku.core;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.List;

public class WebClient extends Client {

	private boolean online;
	private String session;
	
	public WebClient(final Config config) {
		super(config);
		this.online = false;
	}
	
	public void login() throws Exception {
		if (online) {
			logout();
		}
		
		String serverUrl = config.getString(Config.SERVER_URL);
		String extension = config.getString(Config.SERVER_EXTENSION);
		String username = config.getString(Config.SERVER_USERNAME);
		String password = config.getPassword(Config.SERVER_PASSWORD);
		String link = serverUrl + "api/login" + extension + "?u=" + username + "&p=" + password;
		String response = request(link);
		String sign = "Success: ";
		if (response.startsWith(sign)) {
			session = response.substring(sign.length());
			online = true;
			return;
		}
		sign = "Error: ";
		if (response.startsWith(sign)) {
			throw new Exception("Login fail: " + response.substring(sign.length()));
		}
		throw new Exception("Login fail: Invalid api call!");
	}
	
	public void logout() {
		try {
			String serverUrl = config.getString(Config.SERVER_URL);
			String extension = config.getString(Config.SERVER_EXTENSION);
			String link = serverUrl + "api/logout" + extension + "?s=" + session;
			request(link);
		} catch (Exception e) {
		}
		online = false;
		session = "";
	}
	
	public boolean online() {
		return online;
	}
	
	public void clone(String session) {
		this.session = session;
		this.online = true;
	}
	
	public String createGame(final boolean playFirst, final int width, final int height) throws Exception {
		if (!online) {
			login();
		}
		
		String serverUrl = config.getString(Config.SERVER_URL);
		String extension = config.getString(Config.SERVER_EXTENSION);
		String link = serverUrl + "api/create" + extension + "?s=" + session + "&p=" + (playFirst ? "1" : "0") + "&w=" + width + "&h=" + height;
		String response = request(link);
		String sign = "Success: ";
		if (response.startsWith(sign)) {
			String id = response.substring(sign.length());
			return id;
		}
		sign = "Error: ";
		if (response.startsWith(sign)) {
			throw new Exception("Fail to create game: " + response.substring(sign.length()));
		}
		throw new Exception("Fail to create game: Invalid api call!");
	}
	
	public void joinGame(final String gameId) throws Exception {
		if (!online) {
			login();
		}
		
		String serverUrl = config.getString(Config.SERVER_URL);
		String extension = config.getString(Config.SERVER_EXTENSION);
		String link = serverUrl + "api/join" + extension + "?s=" + session + "&g=" + gameId;
		String response = request(link);
		String sign = "Error: ";
		if (response.startsWith(sign)) {
			throw new Exception("Fail to join game: " + response.substring(sign.length()));
		}
		if (!response.equals("Success")) {
			throw new Exception("Fail to join game: Invalid api call!");
		}
	}
	
	public void listGame(final List<String> tag) {
		try {
			tag.clear();
			if (!online) {
				login();
			}

			String serverUrl = config.getString(Config.SERVER_URL);
			String extension = config.getString(Config.SERVER_EXTENSION);
			String link = serverUrl + "api/list" + extension + "?s=" + session;
			String response = request(link);
			String sign = "Success: ";
			if (response.startsWith(sign)) {
				response = response.substring(sign.length()).trim();
				if (response.length() > 0) {
					String[] lines = response.split("\r\n");
					for (int i = 0; i < lines.length; i++) {
						tag.add(lines[i]);
					}
				}
			}
		} catch (Exception e) {
		}
	}
	
	public void makeMove(final Move move) throws Exception {
		String serverUrl = config.getString(Config.SERVER_URL);
		String gameId = config.getString(Config.CURRENT_GAME);
		String extension = config.getString(Config.SERVER_EXTENSION);
		String link = serverUrl + "api/move" + extension + "?s=" + session + "&g=" + gameId + "&r=" + move.getRow() + "&c=" + move.getColumn() + "&p=" + move.getPiece();
		String response = request(link);
		String sign = "Error: ";
		if (response.startsWith(sign)) {
			throw new Exception("Fail to make move: " + response.substring(sign.length()));
		}
		if (!response.equals("Success")) {
			throw new Exception("Fail to make move: Invalid api call!");
		}
	}
	
	public Move lastMove() {
		try {
			if (!online) {
				login();
			}

			String serverUrl = config.getString(Config.SERVER_URL);
			String extension = config.getString(Config.SERVER_EXTENSION);
			String gameId = config.getString(Config.CURRENT_GAME);
			String link = serverUrl + "api/last" + extension + "?s=" + session + "&g=" + gameId;
			String response = request(link);
			String sign = "Success: ";
			if (response.startsWith(sign)) {
				response = response.substring(sign.length()).trim();
				if (response.length() > 0) {
					String[] fields = response.split("\\|");
					byte piece = Board.BLANK;
					try {
						piece = Byte.parseByte(fields[0]);
					} catch (Exception e) {
						piece = Board.BLANK;
					}
					if (piece != Board.BLACK && piece != Board.WHITE) {
						return new Move();
					}
					int row = -1;
					try {
						row = Integer.parseInt(fields[1]);
					} catch (Exception e) {
						row = -1;
					}
					if (row < 0 || row >= config.getInt(Config.BOARD_HEIGHT)) {
						return new Move();
					}
					int column = -1;
					try {
						column = Integer.parseInt(fields[2]);
					} catch (Exception e) {
						column = -1;
					}
					if (column < 0 || column >= config.getInt(Config.BOARD_WIDTH)) {
						return new Move();
					}
					
					return new Move(row, column, piece);
				} else {
					return new Move();
				}
			} else {
				return new Move();
			}
		} catch (Exception e) {
			return new Move();
		}
	}
	
	public void gameState(GameState state) throws Exception {
		if (!online) {
			login();
		}
		
		String serverUrl = config.getString(Config.SERVER_URL);
		String extension = config.getString(Config.SERVER_EXTENSION);
		String gameId = config.getString(Config.CURRENT_GAME);
		String link = serverUrl + "api/state" + extension + "?s=" + session + "&g=" + gameId;
		String response = request(link);
		String sign = "Success: ";
		if (response.startsWith(sign)) {
			response = response.substring(sign.length());
			String[] fields = response.split("\\|");
			state.Joined = fields[0].equals("1");
			state.Cancelled = fields[1].equals("1");
			state.Finished = fields[2].equals("1");
			try {
				state.Victory = Byte.parseByte(fields[3]);
			} catch (Exception e) {
				state.Victory = Board.NO_WIN;
			}
			return;
		}
		sign = "Error: ";
		if (response.startsWith(sign)) {
			throw new Exception("Fail to get game state: " + response.substring(sign.length()));
		}
		throw new Exception("Fail to get game state: Invalid api call!");
		
	}
	
	public void endGame(final boolean finished, final byte victory) throws Exception {
		if (!online) {
			login();
		}
		
		String serverUrl = config.getString(Config.SERVER_URL);
		String extension = config.getString(Config.SERVER_EXTENSION);
		String gameId = config.getString(Config.CURRENT_GAME);
		String link = serverUrl + "api/end" + extension + "?s=" + session + "&g=" + gameId + "&f=" + (finished ? "1" : "0") + "&v=" + victory;
		String response = request(link);
		String sign = "Error: ";
		if (response.startsWith(sign)) {
			throw new Exception("Fail to end game: " + response.substring(sign.length()));
		}
		if (!response.equals("Success")) {
			throw new Exception("Fail to end game: Invalid api call!");
		}
	}
	
	protected String request(String link) throws Exception {
		URL url = new URL(link);
		BufferedReader in = new BufferedReader(new InputStreamReader(url.openStream()));
		String response = "";
		String line = null;
        while ((line = in.readLine()) != null) {
        	if (response.length() > 0) {
        		response += "\r\n";
        	}
        	response += line;
        }
        in.close();

        return response;
	}
	
}
